File: backend/app/schemas/deal.py
"""
Deal schemas for API request/response validation.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import Field, model_validator

from app.core.sanitization import make_sanitized_validator

from .base import BaseSchema, TimestampSchema


class RecentActivityItem(BaseSchema):
    """Compact activity entry embedded in deal responses."""

    action: str
    description: str
    created_at: datetime

    model_config = {"from_attributes": True}


class DealBase(BaseSchema):
    """Base deal schema with common fields."""

    name: str = Field(..., min_length=1, max_length=255)
    deal_type: str = Field(
        ..., pattern="^(acquisition|disposition|development|refinance)$"
    )
    property_id: int | None = None
    assigned_user_id: int | None = None


class DealCreate(DealBase):
    """Schema for creating a new deal."""

    # Initial stage (defaults to initial_review)
    stage: str = Field(
        default="initial_review",
        pattern="^(dead|initial_review|active_review|under_contract|closed|realized)$",
    )

    # Financial Terms
    asking_price: Decimal | None = Field(None, ge=0)
    offer_price: Decimal | None = Field(None, ge=0)

    # Underwriting Metrics
    projected_irr: Decimal | None = Field(None, ge=0, le=1)
    projected_coc: Decimal | None = Field(None, ge=0, le=1)
    projected_equity_multiple: Decimal | None = Field(None, ge=0)
    hold_period_years: int | None = Field(None, ge=1, le=30)

    # Timeline
    initial_contact_date: date | None = None
    target_close_date: date | None = None

    # Source
    source: str | None = Field(None, max_length=100)
    broker_name: str | None = Field(None, max_length=255)
    broker_company: str | None = Field(None, max_length=255)
    competition_level: str | None = Field(None, pattern="^(low|medium|high)$")

    # Notes
    notes: str | None = None
    investment_thesis: str | None = None
    key_risks: str | None = None

    # Metadata
    tags: list[str] | None = None
    priority: str = Field(default="medium", pattern="^(low|medium|high|urgent)$")

    _sanitize = model_validator(mode="before")(
        make_sanitized_validator(
            "name",
            "source",
            "broker_name",
            "broker_company",
            "notes",
            "investment_thesis",
            "key_risks",
            "tags",
        )
    )


class DealUpdate(BaseSchema):
    """Schema for updating a deal. All fields optional except version (optimistic lock)."""

    # Optimistic locking — client must send the version it last read
    version: int = Field(..., description="Current version for optimistic locking")

    name: str | None = Field(None, min_length=1, max_length=255)
    deal_type: str | None = Field(
        None, pattern="^(acquisition|disposition|development|refinance)$"
    )
    property_id: int | None = None
    assigned_user_id: int | None = None

    asking_price: Decimal | None = Field(None, ge=0)
    offer_price: Decimal | None = Field(None, ge=0)
    final_price: Decimal | None = Field(None, ge=0)

    projected_irr: Decimal | None = Field(None, ge=0, le=1)
    projected_coc: Decimal | None = Field(None, ge=0, le=1)
    projected_equity_multiple: Decimal | None = Field(None, ge=0)
    hold_period_years: int | None = Field(None, ge=1, le=30)

    initial_contact_date: date | None = None
    loi_submitted_date: date | None = None
    due_diligence_start: date | None = None
    due_diligence_end: date | None = None
    target_close_date: date | None = None
    actual_close_date: date | None = None

    source: str | None = Field(None, max_length=100)
    broker_name: str | None = Field(None, max_length=255)
    broker_company: str | None = Field(None, max_length=255)
    competition_level: str | None = Field(None, pattern="^(low|medium|high)$")

    notes: str | None = None
    investment_thesis: str | None = None
    key_risks: str | None = None

    tags: list[str] | None = None
    priority: str | None = Field(None, pattern="^(low|medium|high|urgent)$")
    deal_score: int | None = Field(None, ge=0, le=100)

    _sanitize = model_validator(mode="before")(
        make_sanitized_validator(
            "name",
            "source",
            "broker_name",
            "broker_company",
            "notes",
            "investment_thesis",
            "key_risks",
            "tags",
        )
    )


class DealStageUpdate(BaseSchema):
    """Schema for updating deal stage (Kanban movement)."""

    stage: str = Field(
        ...,
        pattern="^(dead|initial_review|active_review|under_contract|closed|realized)$",
    )
    stage_order: int | None = Field(None, ge=0)


class DealResponse(DealBase, TimestampSchema):
    """Schema for deal response."""

    id: int
    version: int = 1
    # Override inherited pattern constraints — response schemas must not re-validate
    # data already stored in the database.  Any value present in the DB is valid here.
    deal_type: str
    name: str
    stage: str
    stage_order: int

    asking_price: Decimal | None = None
    offer_price: Decimal | None = None
    final_price: Decimal | None = None

    projected_irr: Decimal | None = None
    projected_coc: Decimal | None = None
    projected_equity_multiple: Decimal | None = None
    hold_period_years: int | None = None

    initial_contact_date: date | None = None
    loi_submitted_date: date | None = None
    due_diligence_start: date | None = None
    due_diligence_end: date | None = None
    target_close_date: date | None = None
    actual_close_date: date | None = None

    source: str | None = None
    broker_name: str | None = None
    broker_company: str | None = None
    competition_level: str | None = None

    notes: str | None = None
    investment_thesis: str | None = None
    key_risks: str | None = None
    documents: list[dict] | None = None
    activity_log: list[dict] | None = None

    tags: list[str] | None = None
    custom_fields: dict[str, Any] | None = None
    deal_score: int | None = None
    priority: str

    stage_updated_at: datetime | None = None

    # Enrichment fields from extraction data
    total_units: int | None = None
    avg_unit_sf: float | None = None
    current_owner: str | None = None
    last_sale_price_per_unit: float | None = None
    last_sale_date: str | None = None
    t12_return_on_cost: float | None = None
    levered_irr: float | None = None
    levered_moic: float | None = None
    total_equity_commitment: float | None = None
    # Location
    property_city: str | None = None
    submarket: str | None = None
    year_built: int | None = None
    year_renovated: int | None = None
    # Loss factors
    vacancy_rate: float | None = None
    bad_debt_rate: float | None = None
    other_loss_rate: float | None = None
    concessions_rate: float | None = None
    # NOI
    noi_margin: float | None = None
    # Basis
    purchase_price_extracted: float | None = None
    total_acquisition_budget: float | None = None
    basis_per_unit: float | None = None
    # Cap rates
    t12_cap_on_pp: float | None = None
    t3_cap_on_pp: float | None = None
    total_cost_cap_t12: float | None = None
    total_cost_cap_t3: float | None = None
    # Capital
    loan_amount: float | None = None
    lp_equity: float | None = None
    # Exit
    exit_months: float | None = None
    exit_cap_rate: float | None = None
    # Returns
    unlevered_irr: float | None = None
    unlevered_moic: float | None = None
    lp_irr: float | None = None
    lp_moic: float | None = None
    # Map coordinates
    latitude: float | None = None
    longitude: float | None = None

    # Mini activity feed (most recent 1-3 actions)
    recent_activities: list[RecentActivityItem] | None = None


class DealListResponse(BaseSchema):
    """Paginated list of deals."""

    items: list[DealResponse]
    total: int
    page: int
    page_size: int


class DealCursorPaginatedResponse(BaseSchema):
    """Cursor-paginated list of deals."""

    items: list[DealResponse]
    next_cursor: str | None = None
    prev_cursor: str | None = None
    has_more: bool = False
    total: int | None = None


class KanbanBoardResponse(BaseSchema):
    """Kanban board with deals grouped by stage."""

    stages: dict[str, list[DealResponse]]
    total_deals: int
    stage_counts: dict[str, int]


# ── Proforma Returns Response ────────────────────────────────────────────────


class ProformaFieldValue(BaseSchema):
    """A single proforma field with its extracted value."""

    field_name: str
    value_numeric: float | None = None
    value_text: str | None = None


class ProformaFieldGroup(BaseSchema):
    """A group of proforma fields under a category."""

    category: str
    fields: list[ProformaFieldValue]


class ProformaReturnsResponse(BaseSchema):
    """Proforma returns extracted from UW models for a deal."""

    deal_id: int
    deal_name: str
    groups: list[ProformaFieldGroup]
    total: int


# ── Watchlist Status Response ────────────────────────────────────────────────


class WatchlistStatusResponse(BaseSchema):
    """Watchlist status for a deal and the current user."""

    deal_id: int
    is_watched: bool

File: src/lib/api/schemas/deal.ts
/**
 * Zod schemas for deal API responses
 *
 * Validates the raw backend deal JSON (snake_case, numeric strings)
 * and transforms to frontend Deal type. Replaces transformBackendDeal(),
 * parseCityState(), mapBackendStage(), and BackendDealResponse interface.
 */
import { z } from 'zod';
import type { Deal } from '@/types/deal';

// ---------- Helpers ----------

const STAGE_MAP: Record<string, Deal['stage']> = {
  dead: 'dead',
  initial_review: 'initial_review',
  active_review: 'active_review',
  under_contract: 'under_contract',
  closed: 'closed',
  realized: 'realized',
  // Legacy 8-stage backwards compat
  lead: 'initial_review',
  underwriting: 'active_review',
  loi_submitted: 'under_contract',
  due_diligence: 'under_contract',
};

export function mapBackendStage(stage: string): Deal['stage'] {
  return STAGE_MAP[stage] ?? 'initial_review';
}

function parseCityState(name: string): { propertyName: string; city: string; state: string } {
  const match = name.match(/^(.+?)\s*\(([^,]+),\s*([A-Z]{2})\)\s*(?:\(\d{4}\))?$/);
  if (match) {
    return { propertyName: match[1].trim(), city: match[2].trim(), state: match[3].trim() };
  }
  return { propertyName: name, city: '', state: '' };
}

// ---------- Backend deal schema ----------

export const backendDealSchema = z
  .object({
    id: z.number(),
    version: z.number().optional(),
    name: z.string(),
    deal_type: z.string(),
    property_id: z.number().nullable(),
    assigned_user_id: z.number().nullable(),
    stage: z.string(),
    stage_order: z.number(),
    asking_price: z.union([z.string(), z.number()]).nullable(),
    offer_price: z.union([z.string(), z.number()]).nullable(),
    final_price: z.union([z.string(), z.number()]).nullable(),
    projected_irr: z.union([z.string(), z.number()]).nullable(),
    projected_coc: z.union([z.string(), z.number()]).nullable(),
    projected_equity_multiple: z.union([z.string(), z.number()]).nullable(),
    hold_period_years: z.number().nullable(),
    initial_contact_date: z.string().nullable(),
    loi_submitted_date: z.string().nullable().optional(),
    due_diligence_start: z.string().nullable().optional(),
    due_diligence_end: z.string().nullable().optional(),
    target_close_date: z.string().nullable().optional(),
    actual_close_date: z.string().nullable(),
    source: z.string().nullable(),
    broker_name: z.string().nullable(),
    broker_company: z.string().nullable().optional(),
    competition_level: z.string().nullable().optional(),
    notes: z.string().nullable(),
    investment_thesis: z.string().nullable(),
    key_risks: z.string().nullable().optional(),
    documents: z.array(z.unknown()).nullable().optional(),
    activity_log: z.array(z.unknown()).nullable().optional(),
    tags: z.array(z.string()).nullable().optional(),
    custom_fields: z.record(z.string(), z.unknown()).nullable().optional(),
    deal_score: z.number().nullable(),
    priority: z.string().nullable(),
    created_at: z.string(),
    updated_at: z.string(),
    stage_updated_at: z.string().nullable(),
    // Enrichment fields
    total_units: z.number().nullable(),
    avg_unit_sf: z.number().nullable(),
    current_owner: z.string().nullable(),
    last_sale_price_per_unit: z.number().nullable(),
    last_sale_date: z.string().nullable(),
    t12_return_on_cost: z.number().nullable(),
    levered_irr: z.number().nullable().optional(),
    levered_moic: z.number().nullable().optional(),
    total_equity_commitment: z.number().nullable(),
    // New enrichment fields
    property_city: z.string().nullable().optional(),
    submarket: z.string().nullable().optional(),
    year_built: z.number().nullable().optional(),
    year_renovated: z.number().nullable().optional(),
    vacancy_rate: z.number().nullable().optional(),
    bad_debt_rate: z.number().nullable().optional(),
    other_loss_rate: z.number().nullable().optional(),
    concessions_rate: z.number().nullable().optional(),
    noi_margin: z.number().nullable().optional(),
    purchase_price_extracted: z.number().nullable().optional(),
    total_acquisition_budget: z.number().nullable().optional(),
    basis_per_unit: z.number().nullable().optional(),
    t12_cap_on_pp: z.number().nullable().optional(),
    t3_cap_on_pp: z.number().nullable().optional(),
    total_cost_cap_t12: z.number().nullable().optional(),
    total_cost_cap_t3: z.number().nullable().optional(),
    loan_amount: z.number().nullable().optional(),
    lp_equity: z.number().nullable().optional(),
    exit_months: z.number().nullable().optional(),
    exit_cap_rate: z.number().nullable().optional(),
    unlevered_irr: z.number().nullable().optional(),
    unlevered_moic: z.number().nullable().optional(),
    lp_irr: z.number().nullable().optional(),
    lp_moic: z.number().nullable().optional(),
    latitude: z.number().nullable().optional(),
    longitude: z.number().nullable().optional(),
    recent_activities: z.array(z.object({
      action: z.string(),
      description: z.string(),
      created_at: z.string(),
    })).nullable().optional(),
  })
  .transform((d): Deal => {
    const { propertyName, city, state } = parseCityState(d.name);
    const value = d.asking_price
      ? (typeof d.asking_price === 'number' ? d.asking_price : parseFloat(d.asking_price))
      : d.final_price
        ? (typeof d.final_price === 'number' ? d.final_price : parseFloat(d.final_price))
        : 0;
    const now = new Date();
    const created = new Date(d.created_at);
    const daysInPipeline = Math.max(
      0,
      Math.floor((now.getTime() - created.getTime()) / 86400000),
    );
    const stageStart = d.stage_updated_at ? new Date(d.stage_updated_at) : created;
    const daysInStage = Math.max(
      0,
      Math.floor((now.getTime() - stageStart.getTime()) / 86400000),
    );

    return {
      id: String(d.id),
      propertyName: d.name,
      address: { street: propertyName, city, state },
      value,
      capRate: 0,
      stage: mapBackendStage(d.stage),
      daysInStage,
      totalDaysInPipeline: daysInPipeline,
      assignee: '',
      propertyType: d.deal_type || 'acquisition',
      units: d.total_units ?? undefined,
      avgUnitSf: d.avg_unit_sf ?? undefined,
      currentOwner: d.current_owner ?? undefined,
      lastSalePricePerUnit: d.last_sale_price_per_unit ?? undefined,
      lastSaleDate: d.last_sale_date ?? undefined,
      t12ReturnOnCost: d.t12_return_on_cost ?? undefined,
      leveredIrr: d.levered_irr ?? undefined,
      leveredMoic: d.levered_moic ?? undefined,
      totalEquityCommitment: d.total_equity_commitment ?? undefined,
      createdAt: created,
      timeline: [],
      notes: d.notes ?? undefined,
      // New enrichment fields
      propertyCity: d.property_city ?? undefined,
      submarket: d.submarket ?? undefined,
      yearBuilt: d.year_built ?? undefined,
      yearRenovated: d.year_renovated ?? undefined,
      vacancyRate: d.vacancy_rate ?? undefined,
      badDebtRate: d.bad_debt_rate ?? undefined,
      otherLossRate: d.other_loss_rate ?? undefined,
      concessionsRate: d.concessions_rate ?? undefined,
      noiMargin: d.noi_margin ?? undefined,
      purchasePrice: d.purchase_price_extracted ?? undefined,
      totalAcquisitionBudget: d.total_acquisition_budget ?? undefined,
      basisPerUnit: d.basis_per_unit ?? undefined,
      t12CapOnPp: d.t12_cap_on_pp ?? undefined,
      t3CapOnPp: d.t3_cap_on_pp ?? undefined,
      totalCostCapT12: d.total_cost_cap_t12 ?? undefined,
      totalCostCapT3: d.total_cost_cap_t3 ?? undefined,
      loanAmount: d.loan_amount ?? undefined,
      lpEquity: d.lp_equity ?? undefined,
      exitMonths: d.exit_months ?? undefined,
      exitCapRate: d.exit_cap_rate ?? undefined,
      unleveredIrr: d.unlevered_irr ?? undefined,
      unleveredMoic: d.unlevered_moic ?? undefined,
      lpIrr: d.lp_irr ?? undefined,
      lpMoic: d.lp_moic ?? undefined,
      latitude: d.latitude ?? undefined,
      longitude: d.longitude ?? undefined,
      recentActivities: d.recent_activities?.map((a) => ({
        action: a.action,
        description: a.description,
        createdAt: new Date(a.created_at),
      })) ?? undefined,
    };
  });

// ---------- Response schemas ----------

export const dealsListResponseSchema = z.object({
  items: z.array(backendDealSchema),
  total: z.number(),
});
